sw/device/silicon_creator/lib/drivers/hmac.c

```diff
@@ -75,12 +75,22 @@ void hmac_sha256_final(hmac_digest_t *digest) {
   abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_INTR_STATE_REG_OFFSET,
                    reg);
 
-  // Read the digest in reverse to preserve the numerical value.
-  // The least significant word is at HMAC_DIGEST_7_REG_OFFSET.
-  for (size_t i = 0; i < ARRAYSIZE(digest->digest); ++i) {
-    digest->digest[i] =
-        abs_mmio_read32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_DIGEST_7_REG_OFFSET -
-                        (i * sizeof(uint32_t)));
+  uint32_t result, incr;
+  reg = abs_mmio_read32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_CFG_REG_OFFSET);
+  if (bitfield_bit32_read(reg, HMAC_CFG_DIGEST_SWAP_BIT)) {
+    // Big-endian output.
+    result = HMAC_DIGEST_0_REG_OFFSET;
+    incr = sizeof(uint32_t);
+  } else {
+    // Little-endian output.
+    // Note: we rely on 32-bit integer wraparound to cause the result register
+    // index to count down from 7 to 0.
+    result = HMAC_DIGEST_7_REG_OFFSET;
+    incr = (uint32_t) - sizeof(uint32_t);
+  }
+
+  for (size_t i = 0; i < ARRAYSIZE(digest->digest); ++i, result += incr) {
```

Comment:
In the loop, `i` is not subject to wrap-around.  The value `result` (the result register offset) is what counts either up or down (and its type is already `uint32_t`).

I went this route because:
- The output `digest` index always has to count up and always by a known amount.
- We avoid the counter-intuitive wrap-around comparison:
   ```
   for(size_t i= N - 1; i < N; --i) { ... }
   ```
   This particular style requires the reader to realize that when size_t wraps around below zero, it's value becomes close to SIZE_MAX, and thus fails the comparison expression and exits the loop.  It works perfectly well, its just tricky to read.

Reply:
Ah, makes sense, didn't read it carefully enough the first time around. Thanks for the explanation!